sys/net/transport_layer/ng_udp/ng_udp.c

```diff
@@ -0,0 +1,230 @@
+        return 0xffff;
+    }
+    return csum;
+}
+
+void _receive(ng_pktsnip_t *pkt)
+{
+    ng_pktsnip_t *udp;
+    ng_udp_hdr_t *hdr;
+    uint16_t csum;
+    uint32_t port;
+    ng_netreg_entry_t *sendto;
+
+    /* mark UDP header */
+    pkt = ng_pktbuf_start_write(pkt);
+    if (pkt == NULL) {
+        DEBUG("udp: unable to get write access to packet\n");
+        return;
+    }
+    udp = ng_pktbuf_add(pkt, pkt->data, NG_UDP_HDR_LENGTH, NG_NETTYPE_UNDEF);
+    if (udp == NULL) {
+        DEBUG("udp: error marking UDP header, dropping packet\n");
+        ng_pktbuf_release(pkt);
+        return;
+    }
+    hdr = (ng_udp_hdr_t *)udp->data;
+
+    /* validate checksum */
+    csum = _calc_csum(udp, udp->next, pkt);
+    if (hdr->checksum != csum) {
+        DEBUG("udp: received packet with invalid checksum, dropping it\n");
+        ng_pktbuf_release(pkt);
+        return;
+    }
+
+    /* get port (netreg demux context) */
+    port = (uint32_t)NTOHS(hdr->dst_port);
+
+    /* send payload to receivers */
+    sendto = ng_netreg_lookup(NG_NETTYPE_UDP, port);
+    if (sendto == NULL) {
+        DEBUG("udp: unable to forward packet as no one is interested in it\n");
+        ng_pktbuf_release(pkt);
+        return;
+    }
+    ng_pktbuf_hold(pkt, ng_netreg_num(NG_NETTYPE_UDP, port) - 1);
+    while (sendto != NULL) {
+        ng_netapi_receive(sendto->pid, pkt);
+        sendto = ng_netreg_getnext(sendto);
+    }
+}
+
+void _send(ng_pktsnip_t *pkt)
+{
+    ng_udp_hdr_t *hdr;
+    ng_netreg_entry_t *sendto;
+
+    /* get udp snip and hdr */
+    hdr = (ng_udp_hdr_t *)pkt->next->data;
+    /* fill in size field */
+    hdr->length = HTONS(pkt->next->next->size + NG_UDP_HDR_LENGTH);
+
+    /* and forward packet to the network layer */
+    sendto = ng_netreg_lookup(pkt->type, NG_NETREG_DEMUX_CTX_ALL);
+    /* throw away packet if no one is interested */
+    if (sendto == NULL) {
+        DEBUG("udp: cannot send packet because network layer not found\n");
+        ng_pktbuf_release(pkt);
+        return;
+    }
+    /* send packet to network layer */
+    ng_pktbuf_hold(pkt, ng_netreg_num(pkt->type, NG_NETREG_DEMUX_CTX_ALL) - 1);
+    while (sendto != NULL) {
+        ng_netapi_send(sendto->pid, pkt);
+        sendto = ng_netreg_getnext(sendto);
+    }
+}
+
+void *_event_loop(void *arg)
+{
+    (void)arg;
+    msg_t msg;
+    msg_t msg_queue[NG_UDP_MSG_QUEUE_SIZE];
+
+    /* initialize message queue */
+    msg_init_queue(msg_queue, NG_UDP_MSG_QUEUE_SIZE);
+
+    /* dispatch NETAPI messages */
+    while (1) {
+        msg_receive(&msg);
+        switch (msg.type) {
+            case NG_NETAPI_MSG_TYPE_RCV:
+                DEBUG("udp: NG_NETAPI_MSG_TYPE_RCV\n");
+                _receive((ng_pktsnip_t *)msg.content.ptr);
+                break;
+            case NG_NETAPI_MSG_TYPE_SND:
+                DEBUG("udp: NG_NETAPI_MSG_TYPE_SND\n");
+                _send((ng_pktsnip_t *)msg.content.ptr);
+                break;
+            default:
```

Comment:
Needs to at least reply `-ENOTSUP` to `NG_NETAPI_MSG_TYPE_GET` and  `NG_NETAPI_MSG_TYPE_SET`